### PR TITLE
Add store support for ProxySQL

### PR DESCRIPTION
**Pull request by timvaillancourt**
This PR adds support for ProxySQL as a Freno "store" for retrieving MySQL inventory

Freno connects to ProxySQL's admin port _(MySQL protocol)_ to query the `stats.stats_mysql_connection_pool` table to gather MySQL inventory. `stats.stats_mysql_connection_pool` can be queried using [the read-only ProxySQL `stats` user](https://proxysql.com/documentation/global-variables/#admin-stats_credentials), limiting any security exposure to the ProxySQL admin port

ProxySQL backends with the status of `ONLINE` and `SHUNNED_REPLICATION_LAG` will be considered for Freno probes, all other nodes are considered unhealthy and are skipped

**A gotcha worth noting**: `stats.stats_mysql_connection_pool` can return a false-positive `ONLINE` state for servers immediately after a `LOAD MYSQL SERVERS TO RUNTIME` but importantly before the first healthcheck to the server. This is because the `status` of servers in `main.mysql_servers` is assumed to be true after being loaded to runtime

To mitigate this gotcha, nodes that are considered unhealthy are [cached briefly in the `ignoreServerCache`](https://github.com/github/freno/compare/proxysql_v0?expand=1#diff-e4b4c19fcec5c578224c99bd3827c9beR34) and [are ignored for polling](https://github.com/github/freno/pull/129/files#diff-e4b4c19fcec5c578224c99bd3827c9beR105-R109) so that we don't mistakenly consider a node to be healthy before ProxySQL actually health-checks it _(and potentially considers it unhealthy)_

~~cc @shlomi-noach here _(because it won't let me tag you as an official reviewer 😢)_~~

### CODE REVIEW

**Comment by timvaillancourt** [comment]
This repo provides a Docker-based playground for MySQL+ProxySQL+Freno, using the branch from this PR to build Freno: https://github.com/timvaillancourt/proxysql-freno

To test it out:
```
$ git clone https://github.com/timvaillancourt/proxysql-freno
$ cd proxysql-freno
$ make
$ make logs
```

**Comment by shlomi-noach** [comment]
> ProxySQL backends with the status of ONLINE will be considered for Freno probes, all other nodes are considered unhealthy and are skipped

Since I'm not very familiar with ProxySQL: what does `ONLINE` stand for? It's critical that it must mean "the server _should be up and running_". Looking at the docs, I'm not entirely sure.

### Assuming ONLINE actually means "healthy": (skip next section if that's not the case)

Watch out for this trap. What you want to collect are servers that _should be healthy_, not servers that _report in proxy as healthy_.

To explain, let's look at HAProxy: suppose you have 10 replicas, all listed in HAProxy and are all healthy. Great! Now one of them starts lagging. Within a few seconds, HAProxy will notice it's lagging and report it as `DOWN`. `freno` **must keep collecting this server**. Let's see what happens if **not**. Let's think of the following scenario: a high load transition pushes replication lag to the limits. First replica begins lagging. HAProxy marks it as `DOWN`. `freno` now ignores that replica, and only checks for the other `9` replicas -- which are healthy. So `freno` reports `OK` to the transition process. The transition doubles down. Now `9` replicas serve the traffic `10` replicas used to. As result, now another replica starts lagging. Within a few seconds HAProxy marks it as `DOWN`. So `freno` ignores it. So `freno` now only looks at the remaining `8` replicas. And says `OK` to the transition, which doubles down. Now only `8` replicas server a `10` replica's job. You can see where this is going, this loic will exhaust all replicas.

Instead, `freno` should _include_ the `DOWN` replicas. It will double down to run its own `xinetd` check on those replicas to see if they are expected to be down (`HTTP 404`). And it will check for replication lag.

**Comment by timvaillancourt** [comment]
> To explain, let's look at HAProxy: suppose you have 10 replicas, all listed in HAProxy and are all healthy. Great! Now one of them starts lagging. Within a few seconds, HAProxy will notice it's lagging and report it as `DOWN`. `freno` **must keep collecting this server**. Let's see what happens if **not**. Let's think of the following scenario: a high load transition pushes replication lag to the limits. First replica begins lagging. HAProxy marks it as `DOWN`. `freno` now ignores that replica, and only checks for the other `9` replicas -- which are healthy. So `freno` reports `OK` to the transition process. The transition doubles down. Now `9` replicas serve the traffic `10` replicas used to. As result, now another replica starts lagging. Within a few seconds HAProxy marks it as `DOWN`. So `freno` ignores it. So `freno` now only looks at the remaining `8` replicas. And says `OK` to the transition, which doubles down. Now only `8` replicas server a `10` replica's job. You can see where this is going, this loic will exhaust all replicas.

@shlomi-noach hello, I hope you're well! This cascading lag scenario is one I didn't think of, thanks for commenting! 👍 

I built this PR by copying the Vitess store in `pkg/vitess`, so I think that store has the same risk - something I'll follow up on separately in an upcoming PR for tablet health support. It would be great to bounce ideas with you/PlanetScale regarding this offline

Your question has made me realize I need to add requirements for this ProxySQL support in a `pkg/proxysql/README.md` - I've developed this assuming [the ProxySQL monitor module](https://github.com/sysown/proxysql/wiki/Monitor-Module) is enabled. This module checks the health of the backends and dynamically changes the `status` column in `stats.stats_mysql_connection_pool` based on the check results

Currently this PR considers `ONLINE` nodes only, meaning a node that passes all monitor-plugin healthchecks _(eg: connect + ping + replication lag)_. As you pointed out in the scenario you described, this has the risk of allowing cascading replication lag cluster-wide

Thankfully, the monitor module has a state called `SHUNNED_REPLICATION_LAG`, meaning a node that is reachable but has replication lag above [the `max_replication_lag` field set in `mysql_servers`](https://github.com/sysown/proxysql/wiki/Main-(runtime)#mysql_servers). ProxySQL won't send reads to nodes in this state, similar to a `DOWN`ed node under HaProxy

I believe that changing the logic to consider both `ONLINE` and `SHUNNED_REPLICATION_LAG` status would address the scenario you're describing. We would continue to ignore `OFFLINE_SOFT` _(a node that is draining traffic - issued manually)_, `SHUNNED` _(a node ProxySQL cannot connect to or ping)_

Does this solution check out @shlomi-noach?

**Comment by shlomi-noach** [comment]
> I believe that changing the logic to probe servers with either ONLINE or SHUNNED_REPLICATION_LAG status would address the scenario you're describing. We would continue to ignore OFFLINE_SOFT (a node that is draining traffic and won't accept new connections - issued manually) and SHUNNED (a node ProxySQL cannot connect to or ping)

I'm again not very familiar with ProxySQL probe statuses, but this seems to make sense. If a server that is _supposed_ to be serving can be in either `ONLINE` or `SHUNNED_REPLICATION_LAG` then yeah.

> This solution avoids deploying http-based checks, something that isn't common on ProxySQL-based deployments

I'm merely guessing `freno` will continue to run http checks on backend's xinetd, regardless of ProxySQL, right? I think that's expected throughout some infra scripts.

File: pkg/proxysql/client.go [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+		switch server.Status {
+		case "ONLINE":
+			if _, ignore := c.ignoreServerCache.Get(server.Address()); ignore {
+				log.Debugf("found %q in the proxysql ignore-server cache, ignoring ONLINE state for %s", server.Address(), ignoreServerTTL)
+				continue
+			}
+			servers = append(servers, server)
+		case "SHUNNED_REPLICATION_LAG":
+			defer c.ignoreServerCache.Delete(server.Address())
+			servers = append(servers, server)
```
Comment (shlomi-noach) [inline pkg/proxysql/client.go:114]: :+1:

File: pkg/throttle/throttler.go [hunk 2]
```diff
@@ -185,6 +192,15 @@ func (throttler *Throttler) Operate() {
 	}
 }
 
+func (throttler *Throttler) hasProxySQLStores() bool {
```
Comment (shlomi-noach) [inline pkg/throttle/throttler.go:195]: seeing that this function is called by `collectMySQLMetrics()`, and from within a goroutine, should this be locked by a mutex? I fail to remember if  `config.Settings()` can be updated live. If not -- ignore this comment,

Comment (timvaillancourt) [reply pkg/throttle/throttler.go:195]: @shlomi-noach good question, will confirm and potentially add a mutex 👍

Comment (timvaillancourt) [reply pkg/throttle/throttler.go:195]: @shlomi-noach it looks like `hasProxySQLStores()` is called from `NewThrottler` only, eg: https://github.com/github/freno/pull/129/files/1e9e2d7023083f89c1b06ffbd64b8936eafbc271#diff-25a1c9a323af9f356741a7aaa4c7f15bR108

So I don't think `config.Settings()` could be changing at this stage, can you confirm?

Comment (shlomi-noach) [reply pkg/throttle/throttler.go:195]: You're right, my mistake!

Comment (shlomi-noach) [reply pkg/throttle/throttler.go:195]: You're right, my mistake!

**Comment by timvaillancourt** [comment]
> I'm merely guessing `freno` will continue to run http checks on backend's xinetd, regardless of ProxySQL, right? I think that's expected throughout some infra scripts.

@shlomi-noach thanks for all the reviews!

In [my testbed environment](https://github.com/timvaillancourt/proxysql-freno/branches) I am not deploying any HTTP healthcheck server on replicas and it all just works ™️, but I'll follow up on whether or not Freno is sending HTTP checks that are potentially failing

My understanding was HTTP checks was opt-in or in the HaProxy support only. Let's see! In the ProxySQL use case this check would probably be unnecessary

**Comment by timvaillancourt** [comment]
> My understanding was HTTP checks was opt-in or in the HaProxy support only. Let's see!

@shlomi-noach it turns out if the [`HttpCheckPort`](https://github.com/github/freno/blob/master/pkg/config/mysql_config.go#L58) is `<= 0` the HTTP checks are "skipped" 🎉 : https://github.com/github/freno/blob/master/pkg/mysql/mysql_http_check.go#L44-L47

Example:
```
tim@Tims-MacBook-Pro proxysql-freno % curl -s http://localhost:8111/debug/vars|grep httpcheck
"httpcheck.skip": 90,
```

This explains why everything just-works on my ProxySQL/Freno testbed where there are no HTTP check servers. This is great because it lets the user decide if they want HTTP checks

**Review by shlomi-noach — APPROVED** [review]
LGTM

**Pull request merged**